xarray/core/parallelcompat.py

```diff
@@ -0,0 +1,367 @@
+"""
+The code in this module is an experiment in going from N=1 to N=2 parallel computing frameworks in xarray.
+It could later be used as the basis for a public interface allowing any N frameworks to interoperate with xarray,
+but for now it is just a private experiment.
+"""
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Generic, Tuple, Type, TypeVar
+
+import numpy as np
+from typing_extensions import TypeAlias
+
+from xarray.core import indexing, utils
+from xarray.core.pycompat import DuckArrayModule, is_duck_dask_array
+
+T_ChunkManager = TypeVar("T_ChunkManager", bound="ChunkManager")
+T_ChunkedArray = TypeVar("T_ChunkedArray")
+T_Chunks: TypeAlias = Tuple[Tuple[int, ...], ...]
+
+CHUNK_MANAGERS: Dict[str, T_ChunkManager] = {}
+
+
+def _get_chunk_manager(name: str) -> "ChunkManager":
+    if name in CHUNK_MANAGERS:
+        chunkmanager = CHUNK_MANAGERS[name]
+        return chunkmanager()
+    else:
+        raise ImportError(f"ChunkManager {name} has not been defined")
+
+
+class ChunkManager(ABC, Generic[T_ChunkedArray]):
+    """
+    Adapter between a particular parallel computing framework and xarray.
+
+    Attributes
+    ----------
+    array_cls
+        Type of the array class this parallel computing framework provides.
+
+        Parallel frameworks need to provide an array class that supports the array API standard.
+        Used for type checking.
+    """
+
+    array_cls: Type[T_ChunkedArray]
+
+    @abstractmethod
+    def __init__(self):
+        ...
+
+    def is_array_type(self, data: Any) -> bool:
+        return isinstance(data, self.array_cls)
+
+    @abstractmethod
+    def chunks(self, data: T_ChunkedArray) -> T_Chunks:
+
+        ...
+
+    @abstractmethod
+    def from_array(
+        self, data: np.ndarray, chunks: T_Chunks, **kwargs
+    ) -> T_ChunkedArray:
+        ...
+
+    @abstractmethod
+    def rechunk(
+        self, data: T_ChunkedArray, chunks: T_Chunks, **kwargs
+    ) -> T_ChunkedArray:
+        ...
+
+    @abstractmethod
+    def compute(self, data: T_ChunkedArray, **kwargs) -> np.ndarray:
+        ...
+
+    @abstractmethod
+    def apply_gufunc(
+        self,
+        func,
+        signature,
+        *args,
+        axes=None,
+        keepdims=False,
+        output_dtypes=None,
+        vectorize=None,
+        **kwargs,
+    ):
+        """
+        Called inside xarray.apply_ufunc, so must be supplied for vast majority of xarray computations to be supported.
+        """
```

Comment:
By choosing which arguments appear here and not in the dask-specific chunk manager, we are defining what exact specification a general parallel backend must support.